filters/include/pcl/filters/spherical_voxel_grid.h

```diff
@@ -0,0 +1,297 @@
+          return;
+        }
+
+        leaf_size_r_ = r;
+        leaf_size_theta_ = static_cast<float> (M_PI) /  vertical_divisions;
+        leaf_size_phi_ = (2.0f * static_cast<float> (M_PI)) / horizontal_divisions;
+
+        leaf_theta_divisions_ = vertical_divisions;
+        leaf_phi_divisions_ = horizontal_divisions;
+      }
+
+      /** \brief Get the radial size of the leaves
+        */
+      inline float
+      getLeafSizeR () const { return (leaf_size_r_); }
+
+      /** \brief Get the vertical angular size (radians) of the leaves
+        */
+      inline float
+      getLeafSizeVertical () const { return (leaf_size_theta_); }
+
+      /** \brief Get the horizontal angular size (radians) of the leaves
+        */
+      inline float
+      getLeafSizeHorizontal () const { return (leaf_size_phi_); }
+
+      /** \brief Get the number of radial divisions
+        */
+      inline int
+      getDivisionsR () const { return (leaf_r_divisions_); }
+
+      /** \brief Get the number of vertical angular divisions
+        */
+      inline int
+      getDivisionsVertical () const { return (leaf_theta_divisions_); }
+
+      /** \brief Get the number of horizontal angular divisions
+        */
+      inline int
+      getDivisionsHorizontal () const { return (leaf_phi_divisions_); }
+
+      /** \brief Set to true if all fields need to be downsampled, or false if just XYZ.
+        * \param[in] downsample the new value (true/false)
+        */
+      inline void
+      setDownsampleAllData (bool downsample) { downsample_all_data_ = downsample; }
+
+      /** \brief Get the state of the internal downsampling parameter (true if
+        * all fields need to be downsampled, false if just XYZ).
+        */
+      inline bool
+      getDownsampleAllData () const { return (downsample_all_data_); }
+
+      /** \brief Set the minimum number of points required for a voxel to be used.
+        * \param[in] min_points_per_voxel the minimum number of points for required for a voxel to be used
+        */
+      inline void
+      setMinimumPointsNumberPerVoxel (unsigned int min_points_per_voxel) { min_points_per_voxel_ = min_points_per_voxel; }
+
+      /** \brief Return the minimum number of points required for a voxel to be used.
+        */
+      inline unsigned int
+      getMinimumPointsNumberPerVoxel () const { return (min_points_per_voxel_); }
+
+      /** \brief Return the maximum radius of a point from the ceter,
+        * i.e. the radius of the spherical voxel grid (after filtering)
+        */
+      inline float
+      getMaximumRadius () const { return (max_radius_); }
+
+      /** \brief Set the cartesian origin used to build the spherical grid.
+        * \param[in] origin the cartesian origin
+        */
+      inline void
+      setOrigin (const Eigen::Vector3f& origin) { filter_origin_ << origin, 0; }
+
+      /** \brief Get the cartesian origin used to build the spherical grid.
+        */
+      inline Eigen::Vector3f
```

Comment:
But we are returning a result of `head<3>()` function call, this can not be a reference.

Reply:
Then what about changing the getters and setters to Eigen::Vector4f which is in reality what we store?

Reply:
What we store is an implementation detail, I don't think it should influence our getters/setters. So the question is rather what is more convenient for the user: non-homogeneous or homogeneous coordinates? And I have no answer to this :)

Reply:
All our point types are inherently homogenous and we do provide Eigen maps for both 3f and 4f types, from and to them, so I feel having the Vector4f version in the getter/setter places no inconvenience from the user perspective, allowing us to "potentially" save a copy.

(That was a long sentence!)

Anyway, It's not a big thing and since we don't seem to have a clear opinion here I leave it up to @sddavis14 to decide in light of the additional information we provided.

Reply:
Will come back to this issue and make a final decision when time permits. For now, I have fixed the other discussed problems.